net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+	struct fi_eq_cm_entry     entry;
+	uint32_t                  event;
+	int                       rc;
+
+	eq = tm->ftm_pep->fep_listen->pep_res.fpr_eq;
+	rc = fi_eq_read(eq, &event, &entry,
+			(sizeof(entry) + sizeof(struct m0_fab__conn_data)), 0);
+	if (rc >= (int)sizeof(entry)) {
+		if (event == FI_CONNREQ) {
+			memset(&en, 0, sizeof(en));
+			cd = (struct m0_fab__conn_data*)entry.data;
+			libfab_ep_ntop(cd->fcd_netaddr, &en);
+			rc = libfab_fab_ep_find(tm, &en, cd->fcd_straddr, &ep);
+			if (rc == FI_SUCCESS) {
+				rc = libfab_conn_accept(ep, tm, entry.info);
+				if (rc != FI_SUCCESS)
+					M0_LOG(M0_ERROR, 
+					       "Conn accept failed %d", rc);
+			}
+			fi_freeinfo(entry.info);
+		}
+	} else if (rc == -FI_EAVAIL) {
+		memset(&eq_err, 0, sizeof(eq_err));
+		rc = fi_eq_readerr(eq, &eq_err, 0);
+		if (rc != sizeof(eq_err)) {
+			M0_LOG(M0_ERROR, "fi_eq_readerr returns error =%s",
+			       fi_strerror((int) -rc));
+		} else {
+			M0_LOG(M0_ERROR, "fi_eq_readerr provider err no %d:%s",
+				eq_err.prov_errno,
+				fi_eq_strerror(eq, eq_err.prov_errno,
+					       eq_err.err_data, NULL, 0));
+		}
+	}
+	/* For all other events, there is no error info available.
+	   Hence, all such events can be ignored.
+	*/
+	return 0;
+}
+
+/**
+ * Check connection established and shutdown events for transmit endpoint.
+ * All other types of events are ignored.
+ */
+static void libfab_txep_event_check(struct m0_fab__ep *txep,
+				    struct m0_fab__active_ep *aep,
+				    struct m0_fab__tm *tm)
+{
+	uint32_t rc;
+
+	rc = libfab_check_for_event(aep->aep_tx_res.ftr_eq);
+	if (rc == FI_CONNECTED) {
+		libfab_pending_bufs_send(txep);
+		aep->aep_tx_state = FAB_CONNECTED;
+	} else if (rc == FI_SHUTDOWN) {
+		/* Reset and reopen endpoint */
+		libfab_txep_init(aep, tm, txep);
+	}
+	/* All other types of events can be ignored */
+}
+
+/**
+ * Check for completion events on the completion queue for the receive endpoint
+ */
+static void libfab_rxep_comp_read(struct fid_cq *cq, struct m0_fab__ep *ep)
+{
+	struct m0_fab__buf *buf[FAB_MAX_COMP_READ];
+	m0_bindex_t         len[FAB_MAX_COMP_READ];
+	uint64_t            data[FAB_MAX_COMP_READ];
+	int                 i;
+	int                 cnt;
+
+	memset(&buf, 0, sizeof(buf));
+	memset(&len, 0, sizeof(len));
+	if (cq != NULL) {
+		cnt = libfab_check_for_comp(cq, buf, len, data);
+		for (i = 0; i < cnt; i++) {
+			if (buf[i] != NULL) {
+				if (buf[i]->fb_length == 0)
+					buf[i]->fb_length = len[i];
+				buf[i]->fb_ev_ep = ep;
+				libfab_buf_done(buf[i], 0);
+			}
+			if (data[i] != 0)
+				libfab_buf_done((struct m0_fab__buf*)data[i],
+						0);
+		}
+	}
+}
+
+/**
+ * Check for completion events on the completion queue for the transmit endpoint
+ */
+static void libfab_txep_comp_read(struct fid_cq *cq)
+{
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__buf       *buf[FAB_MAX_COMP_READ];
+	int                       i;
+	int                       cnt;
+
+	memset(&buf, 0, sizeof(buf));
```

Comment:
We can avoid this memset for optimization as if on average we can less number of events than 256, and the function libfab_check_for_comp correctly sets the buf[] values.

Reply:
Done